core/web/src/main/java/edu/uci/ics/texera/web/resource/QueryPlanResource.java

```diff
@@ -50,7 +51,68 @@
 public class QueryPlanResource {
     
     public static java.nio.file.Path resultDirectory = Utils.getTexeraHomePath().resolve("query-results");
-    
+
+    /**
+     * @param plan Logical plan to be executed
+     * @return Generic GenericWebResponse object
+     */
+    private JsonNode executeMutipleSinkPlan(MultipleSinkPlan plan)  {
+        HashMap<String, ISink> sinkMap = plan.getSinkMap();
+        ObjectNode response = new ObjectMapper().createObjectNode();
+        HashMap<String, List<Tuple>> executionResult = new HashMap<>();
+        // execute the query plan from all sink operators and collect result
+        for (HashMap.Entry<String, ISink> sinkEntry: sinkMap.entrySet()) {
+
+            ISink sinkOperator = sinkEntry.getValue();
+
+            if (sinkOperator instanceof TupleSink) {
+                TupleSink tupleSink = (TupleSink) sinkOperator;
+                tupleSink.open();
+                List<Tuple> result = tupleSink.collectAllTuples();
+                tupleSink.close();
+                executionResult.put(sinkEntry.getKey(), result);
+
+            } else if (sinkOperator instanceof VisualizationOperator) {
+                VisualizationOperator tupleSink = (VisualizationOperator) sinkOperator;
+                tupleSink.open();
+                List<Tuple> result = tupleSink.collectAllTuples();
+                tupleSink.close();
+                executionResult.put(sinkEntry.getKey(), result);
+
+            } else {
+                Engine.getEngine().evaluate(plan);
+            }
+        }
```

Comment:
From line 64 to 85, the workflow is completely rerun multiple times if there are multiple sink operators. In a future PR, this should be changed, so that: operators who output to multiple other operators should keep a buffer of data so that it can return without computing when requested after the first request.

Reply:
In `LogicalPlan.java`, the `connectOperators` method has already used `OneToNBroadcastConnector` which could cache the result if the output arity is greater than 1.

Reply:
oh okay, I did not know that